### ISSUE TITLE
Portable version database

### ISSUE THREAD

**Issue by iamjaydear** [RandomEngy/VidCoder#166]
I want to run the portable version of VC 1.5.33 to do some tests, so I've put it on a Portable Apps USB drive. When I run it it gives me an error that suggests it is seeing newer VC settings on the hard drive. I don't want it to stuff up my v2.44 settings, so how do I get it to run from the USB drive as an isolated, portable app? I expect a portable app to keep ALL of it's settings on the USB drive.

![portable_error](https://cloud.githubusercontent.com/assets/16139156/21833823/6f4eff3e-d807-11e6-81ff-a09dcc65346a.jpg)

**Comment by RandomEngy** [comment]
v2 portable has the ability to read from VidCoder.sqlite that is right beside the executable. 1.5 always reads from the shared settings in %appdata%\VidCoder .

**Comment by iamjaydear** [comment]
Looks like what I want to do isn't going to be possible. I guess what I don't know is if "version 34 database" belongs to v2.44 or some other 1.5 version? What I'm trying to do is process a sample mpg file using both v2.44 and v1.5.33 on the same PC to see if the audio distortion I've reported in another issue is only present in the mp4 from v2.44. My v2.44 install is not portable and I don't want to stuff it up.

**Comment by RandomEngy** [comment]
Yeah when you use a later version it automatically upgrades the database. You'd need to swap back to the old database file when changing to v1.5.33.

**Comment by iamjaydear** [comment]
Bummer :(

**Comment by RandomEngy** [comment]
No, you just need to switch out the DB files. Or use 2.44 Portable and 1.5.33 installed, then there won't be any conflict.

**Comment by sergeevabc** [comment]
How to make ```VidCoder Portable``` keep all its assets (```ImageCache```, ```Logs```, ```Updates```, and ```Sqlite```) in the same folder?

**Comment by RandomEngy** [comment]
I'm curious, Why do you want to do that?

**Comment by sergeevabc** [comment]
@RandomEngy, let’s recall what [portable app][1] means in the first place:
> A portable application does not leave its files or settings on the host computer or modify the existing system and its configuration. The application does not write to the Windows registry or store its configuration files in the user's profile; instead, it stores its configuration files in the portable directory.

Currently so-called VidCoder Portable edition pollutes the host computer instead of keeping its files in the same folder where ```.exe``` resides. This blatant deceit should be fixed as soon as possible.

[1]: https://en.wikipedia.org/wiki/Portable_application

**Comment by sergeevabc** [comment]
Let’s additionally recall what [deceit][1] means then:
> Purposeful falsehood. When you deceive someone, you give them a false impression, mislead them. If someone pretends to be someone else, that’s an act of deceit, like when a thief pretends to be a valet parking attendant and takes off with your car.

When I see a package with a label Milk on it, a nutrient liquid produced by domesticated mammals (e.g. cows) is expected rather than moonshine or semen, which are also white in colour. When I see a package with a label Portable, a specific behaviour is expected, i.e. no traces left behind the folder of execution ever. There are even large websites dedicated to portable apps only with reasons perfectly [laid out][2].

[1]: https://www.vocabulary.com/dictionary/deceit
[2]: https://www.portablefreeware.com/about.php

**Comment by RandomEngy** [comment]
Look, I get how it would be ideal to leave the system without any traces. It's just that there's a lot of stuff to do in VidCoder and if the files it leaves are tiny and don't cause any actual substantive ill effects I'm going to prioritize other work.

Also, deceit is **purposeful** falsehood. VidCoder started as an installed app and I made a portable version that just happens to still write some temporary files to the machine because I haven't moved them yet. And it's actually more complicated to try and write them to the exe directory because you need to gracefully handle when the directory is read-only or doesn't have write permissions. If we suddenly can't use the folder to store temporary preview images, there either needs to be two code paths from giving up the cache, or some new error messages that tell you that previews don't work and why. Anyway I didn't set out to hoodwink anyone, so the term "deceit" is not fair to apply, and frankly annoying.

Anyway, as I read the list of "portable" rules listed there, VidCoder will _never_ be technically "portable", since it relies on the .NET framework, and I'm sure as heck not going to package in a .NET framework bootstrapper. I'll put moving those logs/etc on the backlog to clean up someday, but I'm going to be more interested in issues that actually improve someone's experience with the program rather than checking all the boxes to comply with a dry technical definition.

**Comment by sergeevabc** [comment]
> As I read the list of "portable" rules listed there, VidCoder will never be technically "portable", since it relies on the .NET framework

Slow down as haste makes waste. Nobody said .NET apps could not be portable, the author said: “my preference is not to list any”. Distinctive properties of portable apps are mentioned earlier. Let me repeat Wikipedia’s eponymous article which gives a common denominator of existing opinions on the subject, or a dictionary definition if you will. To call an app portable is to make sure it keeps its settings in the folder of execution and leaves no file/registry traces on another’s machine.

At present VidCoder Portable keeps its settings in SQLite database inside another’s machine profile (C:\Users\…\Roaming) and a few logs around, therefore it is not portable, simple as that.

> I made a portable version that just happens to still write some temporary files to the machine because I haven't moved them yet. And it's actually more complicated to try and write…

Building a roof is also a complicated task, but a house without a roof is not a house yet. Portable app that pollutes system is not a portable app yet. Seeing things the way you wish instead of how they are is called [wishful thinking][1], if a deceit sounds not fair or annoying to you. 

> But I'm going to be more interested in issues that actually improve someone's experience 

Experience begins with comparing claims and reality. When CNN claims Hong Kong is located in South America, then claims Ukraine is located in Pakistan, then again… they lose confidence and eventually turn into fake news outlet. Do you want to share their “glory”? I believe not, so since your app is not yet portable, either remove it or make it work that way.

![](https://cdni.rt.com/files/news/32/c5/b0/00/cnn.jpg)
![](https://cdni.rt.com/files/news/26/a2/c0/00/ukraine-pakistan-cnn-mistake.jpg)


[1]: https://en.wikipedia.org/wiki/Wishful_thinking

**Comment by RandomEngy** [comment]
Ahh that's right the page also says:

> I say this is an ideal definition because very few programs meet all the above requirements. If I were to stick to all these rules, the database probably won't be very useful, at least to me. And not everybody agrees that all these requirements are equally important or necessary.

According to him, there is room to disagree. Anyway, I'm not interested in dropping everything else to meet your strict technical definition of "portable".

But tell you what. If it matters a lot to you, submit a pull request and I'll take a look at it ASAP.

**Comment by sergeevabc** [comment]
Are you aware of dictionary’s definition of “portable”? Means transportable, movable, easy-to-carry version. from Latin word portare, meaning “to carry.” Which means settings go along with executable, not staying in one computer’s profile and re-created in another’s computer.

**Comment by RandomEngy** [comment]
The settings, queue, pickers, presets are all stored in the sqlite file, which DOES get stored alongside the .exe. Here is the full list of things that don't get carried with it:

1) Cached static image files from loading up sources. These are deleted on closing the program.
2) Updates. These are deleted when the update installs.
3) Log files from previous encodes. Normally you only need if something has gone wrong, and are usually used right after the encode.
4) Backup .sqlite databases of previous versions. Not used at all unless you downgrade your VidCoder version.

Which one of these missing is completely breaking your ability to run VidCoder portable on another computer?

**Comment by sergeevabc** [comment]
> Settings stored in the sqlite file, which DOES get stored alongside the .exe.

Not on my end. VidCoder 3.8 Portable x64 creates ```C:\Users\SomeAccount\AppData\Roaming\VidCoder-Beta\VidCoder.sqlite``` I tried deleting that file, it appears again after another launch.

> Log files from previous encodes.
 
They contain sensitive information about processed files.
It allows third-party to grasp what kind of media activities you’re involved in.

**Comment by RandomEngy** [comment]
Do you have permissions to write to that folder? Is it marked as read-only? VidCoder will fall back to using the appdata folder if it can't make a file beside it. I just launched 3.8 Portable and it made a new VidCoder.sqlite file right beside it.

**Comment by sergeevabc** [comment]
> Do you have permissions to write to that folder? Is it marked as read-only?

I even placed ```VidCoder-3.8-Beta-Portable.exe``` to ```C:\``` and launched it as administrator, the same story.
Then I tried launching it from kinda virtual machine provided by [Sandboxie][1], the same story.
OS Windows 7 x64.

[1]: https://www.sandboxie.com/

**Comment by RandomEngy** [comment]
Hmm. I ran my Win7 VM and put the exe on the desktop and it put the .sqlite file beside it. I tried putting it in C: and it also made the .sqlite file beside it there when running as admin.

One thing that might be tripping it up, have you changed the temporary file folder on your setup? Win+R -> cmd -> echo %temp%

If the temp folder is normal and unchanged, I'd like to see a [Process Monitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) log for the .exe folder to see what is happening with the attempted file write.

**Comment by sergeevabc** [comment]
> One thing that might be tripping it up, have you changed the temporary file folder on your setup? 

Yes. I set TEMP and TMP to ```C:\Temp``` via Environmental Variables long ago.

**Comment by RandomEngy** [comment]
Hmm, so close. I have this bit of code:

    IsPortable = Directory.GetCurrentDirectory().Contains("Temp");

Which would still work in your case. Is that folder named "Temp" exactly or maybe "temp" or "TEMP"? That would cause VidCoder to think it was running as non-portable. Under the "Help" menu do you see a menu option for updates? If so, then I think that's what has happened.

**Comment by sergeevabc** [comment]
> $ echo %TEMP%
> c:\temp
> $ echo %TMP%
> c:\temp

but

> $ dir /b /a:d "c:\t*"
> Temp

**Comment by RandomEngy** [comment]
Okay I'm pretty sure that's why the portable version isn't using the correct location for the sqlite file. I'll have this fixed in the next Beta version.

**Comment by sergeevabc** [comment]
> I'll have this fixed in the next Beta version.

Do you believe [this commit][1] is the best approach possible?
What if user comes from Czech Republic and his temporary folder is ```dočasný```…  

[1]: https://github.com/RandomEngy/VidCoder/commit/52efd8a1e9d1389e83e092bd4f09fa71c836aec1

**Comment by RandomEngy** [comment]
Yes, the code should still work for that.

**Comment by sergeevabc** [comment]
Just checked how 3.9 portable behaves on my end. ```VidCoder.sqlite``` is kept alongside the executable now as expected. However ```C:\Users\username\AppData\Roaming\VidCoder-Beta``` is still polluted with image cache and logs containing sensitive information about processed files — it is unacceptable. Some way around this issue should be found as developers of Foobar2000, MPC-HC, CintaNotes, Ocenaudio, SumatraPDF and other prominent pieces of software did.

**Comment by RandomEngy** [comment]
You could submit a pull request with that fixed. As I've mentioned already, it's on the backlog but there are many items prioritized above it.

**Comment by ilkertezcan** [comment]
[Solution](https://www.portablefreeware.com/forums/viewtopic.php?p=65016&sid=1d6b3323857d83d4092b85bb81b254d6#p65016).

**Comment by sergeevabc** [comment]
@ilkertezcan, have you tried it yourself first? After creating ```VidCoder.exe.config``` with the mentioned content inside and launching the app, I see the following error.
```
<appSettings>
     <add key="SettingsDirectory" value="." />
</appSettings>
```
![](https://images.vfl.ru/ii/1549879512/a22517ea/25345765.png)

**Comment by ilkertezcan** [comment]
@sergeevabc: Could you try install to Visual C++ Redists? Recommended [it](https://pastebin.com/raw/46rfrtBv).
